lib/src/persistent_queue.dart

```diff
@@ -0,0 +1,133 @@
+import 'dart:async';
+import 'dart:collection';
+import 'dart:convert';
+
+import 'package:collection/collection.dart';
+import 'package:flutter/foundation.dart';
+import 'package:matomo_tracker/src/dispatch_settings.dart';
+import 'package:matomo_tracker/src/local_storage/local_storage.dart';
+
+class PersistentQueue extends DelegatingQueue<Map<String, String>> {
+  PersistentQueue._(this._storage, Queue<Map<String, String>> base)
+      : _needsSave = false,
+        super(base);
+
+  static Future<PersistentQueue> load({
+    required LocalStorage storage,
+    required PersistenceFilter onLoadFilter,
+  }) async {
+    final actionsData = await storage.loadActions();
+    final actions = decode(actionsData)..retainWhere(onLoadFilter);
+    final queue = PersistentQueue._(storage, Queue.of(actions));
+    await queue.save();
+    return queue;
+  }
+
+  @visibleForTesting
+  static List<Map<String, String>> decode(String? actionsData) {
+    if (actionsData != null) {
+      return (json.decode(actionsData) as List)
+          .cast<Map<String, dynamic>>()
+          .map<Map<String, String>>((element) => element.cast())
+          .toList();
+    } else {
+      return [];
+    }
+  }
+
+  final LocalStorage _storage;
+  bool _needsSave;
+  Completer<void>? _saveInProgress;
+  @visibleForTesting
+  bool get saveInProgress => _saveInProgress != null;
+
+  @visibleForTesting
+  Future<void> save() {
+    _needsSave = true;
+    if (_saveInProgress == null) {
+      _saveInProgress = Completer<void>();
+      unawaited(
+        Future(() async {
+          try {
+            while (_needsSave) {
+              _needsSave = false;
+              final data = json.encode(toList());
+              await _storage.storeActions(data);
+            }
+            _saveInProgress!.complete();
```

Comment:
From my understanding of the `async` and `await` event loop system in dart, `_saveInProgress` can not be `null` here. I prefere using `!` over `?` if I *know* something can not be `null` because I want the app to "explode" if my *knowing* was wrong. If it would become `null` somehow, using `?` would not terminate the app but might cause a lot of side effects without me necessarily noticing immediately, making bug searching (and noticing) a lot harder. In my opinion having an app that just crashes is better than having an app the seems to be running normaly but behaves not as intended.

In this particular case I want to mention that they changed how `async` and `await` work in the past (in [dart release 2.0.0 on 07.08.2018](https://github.com/dart-lang/sdk/blob/main/CHANGELOG.md#200---2018-08-07)) and if something like that would happen again I would prefere noticing by crashing 😉.

I'm open to change it to `?` though if my argumentation isn't appealing to you.

Reply:
While I do understand your point, my philosophy when coding in Dart is to have the "safest" code at compile time and to not trust a dev at a given point and time but what the compiler says. Hence, I'm against the usage of the `!` operator which goes against this philosophy (And I've got the same concern with cast using `as`).

I can see that your code should be safe and I'm sure that you wouldn't have used the `!` operator if you weren't sure that the code would work. But this might not be the case for another contributor and to keep the codebase consistent I prefer to avoid the usage of `!`. (Another thing I should probably write in a contribution guide)

Reply:
I changed it to `?`.

I noticed that in the `return` statement of the method you didn't request a change. Was that on purpose or did you just miss it? I noticed that using `?` wouldn't be possible since we need to return a future and not a `Future?`, so maybe this could would fit the style of the codebase even more:
```dart
  @visibleForTesting
  Future<void> save() {
    _needsSave = true;
    final effectivSaveInProgress = _saveInProgress ?? Completer<void>();
    if (effectivSaveInProgress != _saveInProgress) {
      _saveInProgress = effectivSaveInProgress;
      unawaited(
        Future(() async {
          try {
            while (_needsSave) {
              _needsSave = false;
              final data = json.encode(toList());
              await _storage.storeActions(data);
            }
            effectivSaveInProgress.complete();
          } catch (error, stackTrace) {
            effectivSaveInProgress.completeError(error, stackTrace);
          } finally {
            _saveInProgress = null;
          }
        }),
      );
    }
    return effectivSaveInProgress.future;
  }
```

On the one hand it avoides `?` and `!`, but I feel like it's harder readable... what do you think?

Reply:
Yeah sorry forgot to comment the return too. You should be able to use the `?` operator if you add the `async` keyword to your method and `await` instead of `return`.